internal/server/server.go

```diff
@@ -88,76 +83,46 @@ func (ts *TerminalServer) Run(ctx context.Context) (err error) {
 	subCtx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	mux := cmux.New(ts.listener)
-	defer mux.Close()
-
 	grpcServer := grpc.NewServer()
-	defer grpcServer.GracefulStop()
+	defer grpcServer.Stop()
 	api.RegisterHostServiceServer(grpcServer, ts)
 	api.RegisterGuestServiceServer(grpcServer, ts)
 
-	// Since we use gRPC-Web with Websockets transport, we need additional wrapping
-	webSocketServer := http.Server{
-		Handler: grpcweb.WrapServer(
-			grpcServer,
-			grpcweb.WithWebsockets(true),
-			grpcweb.WithWebsocketOriginFunc(ts.websocketOriginFunc),
-		),
-	}
-	defer func() {
-		if localErr := webSocketServer.Shutdown(context.Background()); localErr != nil {
-			err = localErr
-		}
-	}()
-
-	webSocketListener := mux.Match(
-		cmux.HTTP1HeaderField("content-type", "application/grpc-web+proto"),
-		cmux.HTTP1HeaderField("content-type", "application/grpc-web+proto"),
-		cmux.HTTP1HeaderField("content-type", "application/grpc-web-text"),
-		cmux.HTTP1HeaderField("content-type", "application/grpc-web-text"),
-		cmux.HTTP1HeaderField("Sec-WebSocket-Protocol", "grpc-websockets"),
+	grpcWebServer := grpcweb.WrapServer(
+		grpcServer,
+		grpcweb.WithWebsockets(true),
+		grpcweb.WithWebsocketOriginFunc(ts.websocketOriginFunc),
 	)
 
 	go func() {
 		defer cancel()
 
-		ts.logger.Infof("starting GuestService gRPC-Web server at %s", webSocketListener.Addr().String())
-
-		if err := webSocketServer.Serve(webSocketListener); err != nil {
-			if !errors.Is(err, http.ErrServerClosed) {
-				ts.logger.Warnf("GuestService gRPC-Web server failed: %v", err)
-			}
+		server := http.Server{
+			Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				contentType := r.Header.Get("content-type")
+				switch {
+				case strings.HasPrefix(contentType, "application/grpc-web"):
+					grpcWebServer.ServeHTTP(w, r)
+				case strings.HasPrefix(contentType, "application/grpc"):
```

Comment:
I think we also need to match by `Sec-WebSocket-Protocol` headers here.

Reply:
Tests are passing though. 🤔

Reply:
They work because we don't emulate Improbable's [experimental WebSocket protocol](https://grpc.io/blog/state-of-grpc-web/#f19) and connect as a guest using a normal gRPC protocol:

https://github.com/cirruslabs/terminal/blob/cddc4a522d97890b417519cc9226eddd9724fdbb/internal/server/e2e_test.go#L75-L76